### PR TITLE
Implement Instance Groups

### PR DESCRIPTION
**Pull request by Jamie-BitFlight**
## what
* Create groups of static EC2 instances (e.g. like a `StatefulSet` in kubernetes)

## why
* Stateful applications like databases (kafka) do not belong in AutoScaling Groups

### CODE REVIEW

File: outputs.tf [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+output "security_group_ids" {
+  description = "ID on the new AWS Security Group associated with creating instance"
+  value       = "${compact(concat(list(var.create_default_security_group == "true" ? join("", aws_security_group.default.*.id) : ""), var.security_groups))}"
+}
+
+output "role" {
+  description = "Name of AWS IAM Role associated with creating instance"
+  value       = "${join(",", aws_iam_role.default.*.name)}"
```
Comment (osterman) [inline outputs.tf]: I think we should return a `list` for all outputs since this module is for N servers

File: outputs.tf [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+output "alarm" {
+  description = "CloudWatch Alarm ID"
+  value       = "${join(",", aws_cloudwatch_metric_alarm.default.*.id)}"
```
Comment (osterman) [inline outputs.tf]: return a `list`?

File: outputs.tf [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+output "ebs_ids" {
+  description = "ID of EBSs"
+  value       = "${join(",", aws_ebs_volume.default.*.id)}"
+}
+
+output "primary_network_interface_id" {
+  description = "ID of the instance's primary network interface"
+  value       = "${join(",", aws_instance.default.*.primary_network_interface_id)}"
```
Comment (osterman) [inline outputs.tf]: return a `list`?

File: outputs.tf [hunk 4]
```diff
@@ -0,0 +1,87 @@
+output "public_ips" {
+  description = "List of Public IPs of instances (or EIP)"
+  value       = "${distinct(compact(concat(aws_eip.default.*.public_ip,aws_instance.default.*.public_ip, aws_eip.additional.*.public_ip, list() )))}"
+}
+
+output "private_ip" {
+  description = "Private IPs of instances"
+  value       = "${split(",",join(",", aws_instance.default.*.private_ip))}"
```
Comment (osterman) [inline outputs.tf]: Why do we use the `split-join` pattern here an not else where?

File: README.md [hunk 5]
```diff
@@ -1 +1,225 @@
-# terraform-aws-ec2-instance-group
\ No newline at end of file
+# terraform-aws-ec2-instance-group [![Build Status](https://travis-ci.org/cloudposse/terraform-aws-ec2-instance-group.svg?branch=master)](https://travis-ci.org/cloudposse/terraform-aws-ec2-instance-group)
+
+Terraform Module for providing N general purpose EC2 hosts.
```
Comment (osterman) [inline README.md]: Maybe add a note like:

> If you only need to provision a single EC2 instance, consider using the [`terraform-aws-ec2-instance`](https://github.com/cloudposse/terraform-aws-ec2-instance) module instead.

Comment (osterman) [reply README.md]: Let's also add a statement about rationale here. 
```
**IMPORTANT** This module by-design does not provision an AutoScaling group. It was designed to provision a discrete number of instances suitable for running stateful services such as databases (E.g. Kafka, Redis, etc). 
```

File: README.md [hunk 6]
```diff
@@ -1 +1,230 @@
...
+
+Note: add `${var.ssh_key_pair}` private key to the `ssh agent`.
+
+Include this repository as a module in your existing terraform code.
+
+
+### Simple example:
+
+```hcl
+module "instance" {
```
Comment (aknysh) [inline README.md]: instances?

File: README.md [hunk 7]
```diff
@@ -1 +1,230 @@
...
+  source                      = "git::https://github.com/cloudposse/terraform-aws-ec2-instance-group.git?ref=master"
+  namespace                   = "${var.namespace}"
```
Comment (aknysh) [inline README.md]: let's use concrete `namespace`, `stage`, `name` and other attributes in `README` examples.
easier to read and provides a complete example

File: README.md [hunk 8]
```diff
@@ -1 +1,230 @@
...
+ 2. **Clone** the project to your own machine
+ 3. **Commit** changes to your own branch
+ 4. **Push** your work back up to your fork
+ 5. Submit a **Pull request** so that we can review your changes
+
+**NOTE:** Be sure to merge the latest from "upstream" before making a pull request!
+
+## License
+
+[APACHE 2.0](LICENSE) © 2016-2017 [Cloud Posse, LLC](https://cloudposse.com)
```
Comment (aknysh) [inline README.md]: 2018

File: main.tf [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    principals {
+      type        = "Service"
+      identifiers = ["ec2.amazonaws.com"]
+    }
+
+    effect = "Allow"
+  }
+}
+
+data "aws_ami" "default" {
```
Comment (aknysh) [inline main.tf]: why use these `aws_ami` lookups in `main.tf`?
they will be different for every project.
it's ok to put it in the examples.

File: main.tf [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+data "aws_ami" "info" {
+  filter {
+    name   = "image-id"
+    values = ["${local.ami}"]
+  }
+}
+
+# Apply the tf_label module for this resource
```
Comment (aknysh) [inline main.tf]: terraform-terraform-label

File: outputs.tf [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+  # Split/Join used here so that the replace can be used inbetween.
+  ip_dns_list = "${split(",", replace(join(",", distinct(compact(concat(aws_eip.default.*.public_ip,aws_instance.default.*.public_ip, aws_eip.additional.*.public_ip, list() )))), ".", "-"))}"
+  dns_names   = "${formatlist("%v.${var.region == "us-east-1" ? "compute-1" : "${var.region}.compute"}.amazonaws.com", distinct(compact(local.ip_dns_list)))}"
+}
+
+output "public_dns" {
+  value = "${local.dns_names}"
+}
+
+output "id" {
```
Comment (aknysh) [inline outputs.tf]: ids

File: outputs.tf [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+  value       = "${aws_instance.default.*.id}"
+}
+
+output "aws_key_pair_name" {
+  description = "Name of AWS key pair"
+  value       = "${signum(length(var.ssh_key_pair)) == 1 ? var.ssh_key_pair : "${var.generate_ssh_key_pair == "true" ? module.ssh_key_pair.key_name : ""}"}"
+}
+
+output "new_ssh_keypair_generated" {
+  value = "${signum(length(var.ssh_key_pair)) == 1 ? "false" : "true" }"
```
Comment (aknysh) [inline outputs.tf]: add `description`

File: outputs.tf [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+  description = "ID on the new AWS Security Group associated with creating instance"
+  value       = "${compact(concat(list(var.create_default_security_group == "true" ? join("", aws_security_group.default.*.id) : ""), var.security_groups))}"
+}
+
+output "roles" {
+  description = "Name of AWS IAM Role associated with creating instance"
+  value       = "${compact(aws_iam_role.default.*.name)}"
+}
+
+output "alarms" {
```
Comment (aknysh) [inline outputs.tf]: `alarm_ids`

File: outputs.tf [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+  description = "Path where SSH key pair was created (if applicable)"
+  value       = "${local.ssh_key_pair_path}/${module.ssh_key_pair.key_name}.pem"
+}
+
+output "security_group_ids" {
+  description = "ID on the new AWS Security Group associated with creating instance"
+  value       = "${compact(concat(list(var.create_default_security_group == "true" ? join("", aws_security_group.default.*.id) : ""), var.security_groups))}"
+}
+
+output "roles" {
```
Comment (aknysh) [inline outputs.tf]: `role_names`

File: outputs.tf [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+  description = "IDs of the instance's primary network interface"
+  value       = "${aws_instance.default.*.primary_network_interface_id}"
+}
+
+output "network_interface_ids" {
+  description = "IDs of the network interface that was created with the instance"
+  value       = "${aws_instance.default.*.network_interface_id}"
+}
+
+output "eips_per_instance" {
```
Comment (aknysh) [inline outputs.tf]: `eip_per_instance_count`

File: user-data/user_data.sh [hunk 16]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
```
Comment (aknysh) [inline user-data/user_data.sh]: `user_data.sh` is project-specific.
should be configurable

File: variables.tf [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+  description = "List of allowed ingress ports"
+  default     = []
+}
+
+variable "subnet" {
+  description = "VPC Subnet ID the instance is launched in"
+}
+
+variable "namespace" {
+  description = "Namespace (e.g. `cp` or `cloudposse`) - required for `tf_label` module"
```
Comment (aknysh) [inline variables.tf]: terraform-terraform-label

File: variables.tf [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+}
+
+variable "stage" {
+  description = "Stage (e.g. `prod`, `dev`, `staging` - required for `tf_label` module"
```
Comment (aknysh) [inline variables.tf]: terraform-terraform-label

File: variables.tf [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+}
+
+variable "name" {
+  description = "Name  (e.g. `bastion` or `db`) - required for `tf_label` module"
```
Comment (aknysh) [inline variables.tf]: terraform-terraform-label

File: variables.tf [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+  description = "Enable EC2 Instance Termination Protection"
+  default     = "false"
+}
+
+variable "monitoring" {
+  description = "Launched EC2 instance will have detailed monitoring enabled"
+  default     = "true"
+}
+
+variable "private_ip" {
```
Comment (aknysh) [inline variables.tf]: is this the same private IP address for all instances?
if we create more than one instance, will it work?
probably need to provide a list of IPs.

File: variables.tf [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+  description = "Size of the root volume in gigabytes"
+  default     = "10"
+}
+
+variable "root_iops" {
+  description = "Amount of provisioned IOPS. This must be set if root_volume_type is set to `io1`"
+  default     = "0"
+}
+
+variable "ebs_device_name" {
```
Comment (aknysh) [inline variables.tf]: if it's a `list`, then name it `ebs_device_names` and update the description

File: README.md [hunk 22]
```diff
@@ -1 +1,230 @@
...
+  instance_count              = "3"
+}
+```
+
+### Additional complete working example with variations of how to use the module
+In /examples directory
+
+This module depends on these modules:
+
+* [terraform-null-label](https://github.com/cloudposse/terraform-null-label)
```
Comment (aknysh) [inline README.md]: terraform-terraform-label

File: examples/main.tf [hunk 23]
```diff
@@ -0,0 +1,144 @@
...
+
+data "aws_subnet_ids" "all" {
+  vpc_id = "${data.aws_vpc.default.id}"
+}
+
+provider "aws" {
+  region = "us-east-1"
+}
+
+module "zero_servers" {
```
[the remark this answers is not in the document]
Comment (Jamie-BitFlight) [reply examples/main.tf]: Yeah. Example shows it working in multiple test cases. So when it is run you can see what outputs are available from each case.

File: main.tf (hunk 9 above)
Comment (Jamie-BitFlight) [reply main.tf]: I missed that :-P

File: user-data/user_data.sh (hunk 16 above)
Comment (Jamie-BitFlight) [reply user-data/user_data.sh]: It is configurable via the "user_data" variable.

File: README.md [hunk 24]
```diff
@@ -1 +1,230 @@
...
+  subnet                      = "${var.subnet}"
+
+  instance_count              = "3"
+}
+```
+
+### Example with additional volumes and EIP
+
+```hcl
+module "kafka_instance" {
```
[the remark this answers is not in the document]
Comment (Jamie-BitFlight) [reply README.md]: This was the example from the terraform_aws_ec2_instance so I left it there.

**Comment by Jamie-BitFlight** [comment]
Thank you @aknysh I have addressed those comments now.

**Comment by osterman** [comment]
@aknysh for final approval

**Pull request merged**